### PR TITLE
[core] Set OMP_NUM_THREADS to `num_cpus` required by task/actors by default

### LINKED ISSUE #24012: [Core] OMP_NUM_THREADS change
### Description

It would be great if OMP_NUM_THREADS only changed within the context of a ray.get, ray.wait, tune.run, etc call so it doesn't affect other code.

### Use case

We just upgraded from ray from 1.8.0 to 1.11.0, and the change where importing ray sets OMP_NUM_THREADS to 1 has caused many issues for us. Examples:

1. A model pipeline that runs on a single machine might use ray to parallelize some data work then use xgboost or pytorch for thread-based model training or do something else that needs more than one thread.
2. A distributed pipeline might use ray.tune to optimize xgb hyperparameters where different trials happen on different nodes. Here, resources per trial would be something like `{cpu: os.cpu_count()}`.
3. Similarly, we use ray core to do distributed cross validation with xgb models. In this case, we wrap our training functions with `ray.remote(num_cpus=os.cpu_count())` and expect them to use all cores on whichever node they're on.
4. In a single function call, you may want to limit threads to 1 per worker (e.g., parallelizing some numpy code) then set them back to max (e.g., calling a function that does its own threading). This isn't an issue caused by the change, but we've had to do this before.

For many of our use cases, globally changing OMP_NUM_THREADS is a huge side-effect. My (probably uninformed) suggestion would be to make threads a resource like CPUs, GPUs, etc.

### LINKED ISSUE #25837: [Core] Changing the OMP_NUM_THREADS with num_cpus
### Description

I was trying to use Ray Core for parallelising my task over nodes in a cluster. However my task were build in C++ and they are optimized by openmp. The problem is that even if I set num_cpus to K, the OMP_NUM_THREADS remain to be 1. So, one feature what we can do is if a user sets num_cpus to K, OMP_NUM_THREADS should be increased to K.

### Use case

People do assume that if we change num_cpus to K, the omp_num_threads to be K too. But it remain to be 1. Could we add something to make sure that when num_cpus to K, the omp_num_threads should also be change to K.

### PR DESCRIPTION
**Pull request by rickyyx**
Signed-off-by: rickyyx <rickyx@anyscale.com>

## Why are these changes needed?
Ray currently sets `OMP_NUM_THREADS=1` when the environ variable is not set. 
This PR:
- Sets `OMP_NUM_THREADS` to the number of cpus assigned to the worker that runs a task before running, and reset it after running. 
  - If `num_cpus` is a fractional smaller than 1, it will set `OMP_NUM_THREADS` to 1. 
- Doesn't override `OMP_NUM_THREADS` if it's already being specified in runtime env or through os.environ. 

## Related issue number
Closes #24012 
Closes #25837

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: doc/source/ray-core/configure.rst [hunk 1]
```diff
@@ -57,11 +57,13 @@ If using the command line, connect to the Ray cluster as follow:
...
 
 .. note::
-    Ray sets the environment variable ``OMP_NUM_THREADS=1`` by default. This is done
-    to avoid performance degradation with many workers (issue #6998). You can
-    override this by explicitly setting ``OMP_NUM_THREADS``. ``OMP_NUM_THREADS`` is commonly
-    used in numpy, PyTorch, and Tensorflow to perform multi-threaded linear algebra.
-    In multi-worker setting, we want one thread per worker instead of many threads
+    Ray sets the environment variable ``OMP_NUM_THREADS=<num_cpus>`` if ``num_cpus`` is set on 
+    the task/actor via :ref:`ray.remote() <ray-remote-ref>` and :ref:`.options() <ray-options-ref>`.
+    Or Ray sets ``OMP_NUM_THREADS=1`` if ``num_cpus`` is not specified; this
```
Comment (ericl) [inline doc/source/ray-core/configure.rst:62]:
```suggestion
    Ray sets ``OMP_NUM_THREADS=1`` if ``num_cpus`` is not specified; this 
```

File: doc/source/serve/scaling-and-resource-allocation.md [hunk 2]
```diff
@@ -152,7 +152,7 @@ Below is the full list of supported options in `ray_actor_options`; please see t
 
 Deep learning models like PyTorch and Tensorflow often use multithreading when performing inference.
 The number of CPUs they use is controlled by the `OMP_NUM_THREADS` environment variable.
-To [avoid contention](omp-num-thread-note), Ray sets `OMP_NUM_THREADS=1` by default because Ray tasks and actors use a single CPU by default.
+Ray sets `OMP_NUM_THREADS=<num_cpus>` by default. To [avoid contention](omp-num-thread-note), Ray sets `OMP_NUM_THREADS=1` if `num_cpus` is not specified on the tasks/actors, to reduce contention between actors/tasks which run in a single thread.
 If you *do* want to enable this parallelism in your Serve deployment, just set `OMP_NUM_THREADS` to the desired value either when starting Ray or in your function/class definition:
```
Comment (simon-mo) [inline doc/source/serve/scaling-and-resource-allocation.md:156]:
```suggestion
If you *do* want to enable this parallelism in your Serve deployment, just set `num_cpus` (recommended) to the desired value, or manually set the `OMP_NUM_THREADS` environment variable when starting Ray or in your function/class definition.
```

**Review by simon-mo — APPROVED** [review]

File: python/ray/tests/test_basic.py [hunk 3]
```diff
@@ -120,10 +120,86 @@ def f():
...
+    ###########################
+    @ray.remote
+    def g():
+        return os.getpid(), os.environ.get("OMP_NUM_THREADS")
+
+    # Set to 1
+    pid1, omp_num_threads = ray.get(g.remote())
+    assert omp_num_threads == "1"
+    # Set to 2
+    pid2, omp_num_threads = ray.get(g.options(num_cpus=2).remote())
```
Comment (ericl) [inline python/ray/tests/test_basic.py:192]: Nice test.

**Review by ericl — APPROVED** [review]

**Review by sihanwang41 — APPROVED** [review]
awesome!

**Pull request merged**